[Bug] [Doris] The "SQL generation" function cannot be used in version 0.7.4 "Metadata Center" connection to the Doris data warehouse

### Search before asking

- [X] I had searched in the [issues](https://github.com/DataLinkDC/dinky/issues?q=is%3Aissue) and found no similar issues.


### What happened

<img width="808" alt="image" src="https://github.com/DataLinkDC/dinky/assets/25876796/79b7d2a2-ef5d-46a5-956f-a391cdf6bd53">
Caused by: java.lang.StackOverflowError
 at com.dlink.metadata.driver.AbstractJdbcDriver.listColumnsSortByPK(AbstractJdbcDriver.java:390) ~[dlink-metadata-base-0.7.4.jar:?]
 at com.dlink.metadata.driver.DorisDriver.listColumns(DorisDriver.java:187) ~[dlink-metadata-doris-0.7.4.jar:?]

<img width="714" alt="image" src="https://github.com/DataLinkDC/dinky/assets/25876796/87133334-26f6-4340-aa6d-c13fd5e3d23a">


### What you expected to happen

可以正常自动生成SQL

### How to reproduce

问题版本0.7.4

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dlink-metadata/dlink-metadata-base/src/main/java/com/dlink/metadata/driver/AbstractJdbcDriver.java
+++ b/dlink-metadata/dlink-metadata-base/src/main/java/com/dlink/metadata/driver/AbstractJdbcDriver.java
@@ -386,13 +386,6 @@ public abstract class AbstractJdbcDriver extends AbstractDriver {
     }
 
     @Override
-    public List<Column> listColumnsSortByPK(String schemaName, String tableName) {
-        List<Column> columnList = listColumns(schemaName, tableName);
-        columnList.sort(Comparator.comparing(Column::isKeyFlag).reversed());
-        return columnList;
-    }
-
-    @Override
     public boolean createTable(Table table) throws Exception {
         String sql = getCreateTableSql(table).replaceAll("\r\n", " ");
         if (Asserts.isNotNull(sql)) {

--- a/dlink-metadata/dlink-metadata-base/src/main/java/com/dlink/metadata/driver/Driver.java
+++ b/dlink-metadata/dlink-metadata-base/src/main/java/com/dlink/metadata/driver/Driver.java
@@ -153,8 +153,6 @@ public interface Driver extends AutoCloseable {
 
     List<Column> listColumns(String schemaName, String tableName);
 
-    List<Column> listColumnsSortByPK(String schemaName, String tableName);
-
     List<Schema> getSchemasAndTables();
 
     List<Table> getTablesAndColumns(String schemaName);

--- a/dlink-metadata/dlink-metadata-doris/src/main/java/com/dlink/metadata/driver/DorisDriver.java
+++ b/dlink-metadata/dlink-metadata-doris/src/main/java/com/dlink/metadata/driver/DorisDriver.java
@@ -19,13 +19,6 @@
 
 package com.dlink.metadata.driver;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
-import cn.hutool.core.text.CharSequenceUtil;
 import com.dlink.metadata.convert.DorisTypeConvert;
 import com.dlink.metadata.convert.ITypeConvert;
 import com.dlink.metadata.query.DorisQuery;
@@ -37,6 +30,15 @@ import com.dlink.process.context.ProcessContextHolder;
 import com.dlink.process.model.ProcessEntity;
 import com.dlink.utils.LogUtil;
 import com.dlink.utils.SqlUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import cn.hutool.core.text.CharSequenceUtil;
 
 public class DorisDriver extends AbstractJdbcDriver {
 
@@ -133,11 +135,13 @@ public class DorisDriver extends AbstractJdbcDriver {
 
     @Override
     public String getCreateTableSql(Table table) {
-        List<String> dorisTypes = Arrays.asList("BOOLEAN", "TINYINT", "SMALLINT", "SMALLINT", "INT", "BIGINT", "LARGEINT", "FLOAT", "DOUBLE", "DECIMAL", "DATE", "DATETIME", "CHAR", "VARCHAR", "TEXT", "TIMESTAMP", "STRING");
+        List<String> dorisTypes = Arrays.asList("BOOLEAN", "TINYINT", "SMALLINT", "SMALLINT", "INT", "BIGINT",
+                "LARGEINT", "FLOAT", "DOUBLE", "DECIMAL", "DATE", "DATETIME", "CHAR", "VARCHAR", "TEXT", "TIMESTAMP",
+                "STRING");
         StringBuilder keyBuffer = new StringBuilder();
         StringBuilder ddlBuffer = new StringBuilder();
         ddlBuffer.append("CREATE TABLE IF NOT EXISTS ").append(table.getSchema()).append(".").append(table.getName())
-            .append(" (").append(System.lineSeparator());
+                .append(" (").append(System.lineSeparator());
         for (int i = 0; i < table.getColumns().size(); i++) {
             Column columnInfo = table.getColumns().get(i);
             String cType = columnInfo.getType().split(" ")[0].toUpperCase();
@@ -146,27 +150,29 @@ public class DorisDriver extends AbstractJdbcDriver {
                 cType = columnInfo.getJavaType().getFlinkType().toUpperCase();
                 if (!dorisTypes.contains(cType)) {
                     logger.error("doris does not support {} type", columnInfo.getType());
-                    return  "";
+                    return "";
                 }
             }
             if (cType.equalsIgnoreCase("TIMESTAMP")) {
                 ddlBuffer.append("DATETIME");
             } else if (columnInfo.getType().equalsIgnoreCase("TEXT")) {
                 ddlBuffer.append("STRING");
-            }else{
+            } else {
                 ddlBuffer.append(cType);
             }
-            if(columnInfo.getLength()!=null &&columnInfo.getLength()>0 ) {
-                ddlBuffer.append("(").append(cType.equalsIgnoreCase("VARCHAR") ?columnInfo.getLength() * 3 : columnInfo.getLength()).append(")");
+            if (columnInfo.getLength() != null && columnInfo.getLength() > 0) {
+                ddlBuffer.append("(")
+                        .append(cType.equalsIgnoreCase("VARCHAR") ? columnInfo.getLength() * 3 : columnInfo.getLength())
+                        .append(")");
             }
-            if (columnInfo.getComment()!=null) {
+            if (columnInfo.getComment() != null) {
                 ddlBuffer.append(" COMMENT '").append(columnInfo.getComment()).append("'");
             }
             if (i < table.getColumns().size() - 1) {
                 ddlBuffer.append(",");
             }
             ddlBuffer.append(System.lineSeparator());
-            if(columnInfo.isKeyFlag()){
+            if (columnInfo.isKeyFlag()) {
                 keyBuffer.append(columnInfo.getName()).append(",");
             }
 
@@ -175,7 +181,8 @@ public class DorisDriver extends AbstractJdbcDriver {
         String primaryKeys = keyBuffer.substring(0, keyBuffer.length() - 1);
         ddlBuffer.append(") UNIQUE KEY (").append(primaryKeys).append(")").append(System.lineSeparator());
         ddlBuffer.append("COMMENT '").append(table.getComment()).append("'");
-        ddlBuffer.append(" DISTRIBUTED BY HASH (").append(primaryKeys).append(") BUCKETS AUTO").append(System.lineSeparator());
+        ddlBuffer.append(" DISTRIBUTED BY HASH (").append(primaryKeys).append(") BUCKETS AUTO")
+                .append(System.lineSeparator());
         //
         ddlBuffer.append(" PROPERTIES ( \"replication_allocation\" = \"tag.location.default: 3\")");
         return ddlBuffer.toString();
@@ -184,6 +191,8 @@ public class DorisDriver extends AbstractJdbcDriver {
     @Override
     public List<Column> listColumns(String schemaName, String tableName) {
         // Doris 中声明为 Key 的列（可能是多个）必须顺序声明在建表语句头部，因此按 Key 对列重新排序
-        return listColumnsSortByPK(schemaName, tableName);
+        List<Column> columnList = super.listColumns(schemaName, tableName);
+        columnList.sort(Comparator.comparing(Column::isKeyFlag).reversed());
+        return columnList;
     }
 }

--- a/dlink-metadata/dlink-metadata-starrocks/src/main/java/com/dlink/metadata/driver/StarRocksDriver.java
+++ b/dlink-metadata/dlink-metadata-starrocks/src/main/java/com/dlink/metadata/driver/StarRocksDriver.java
@@ -29,6 +29,7 @@ import com.dlink.utils.LogUtil;
 import com.dlink.utils.SqlUtil;
 
 import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -112,6 +113,8 @@ public class StarRocksDriver extends AbstractJdbcDriver {
     @Override
     public List<Column> listColumns(String schemaName, String tableName) {
         // StarRocks 中声明为 Key 的列（可能是多个）必须顺序声明在建表语句头部，因此按 Key 对列重新排序
-        return listColumnsSortByPK(schemaName, tableName);
+        List<Column> columnList = super.listColumns(schemaName, tableName);
+        columnList.sort(Comparator.comparing(Column::isKeyFlag).reversed());
+        return columnList;
     }
 }

